File: templates/problems/02_two_pointers/04_container_most_water/python/solution.py
def solve_max_area(height: list[int]) -> int:
    """Find the maximum area between two lines using two pointers.

    Time: O(n), Space: O(1)
    """
    l, r = 0, len(height) - 1
    best = 0
    while l < r:
        h = min(height[l], height[r])
        area = h * (r - l)
        if area > best:
            best = area
        if height[l] < height[r]:
            l += 1
        else:
            r -= 1
    return best

File: templates/problems/02_two_pointers/04_container_most_water/typescript/solution.ts
/**
 * Find the maximum area between two lines using two pointers.
 * Time: O(n), Space: O(1)
 */
export function solveMaxArea(height: number[]): number {
    let l = 0;
    let r = height.length - 1;
    let best = 0;
    while (l < r) {
        const h = Math.min(height[l], height[r]);
        const area = h * (r - l);
        if (area > best) {
            best = area;
        }
        if (height[l] < height[r]) {
            l++;
        } else {
            r--;
        }
    }
    return best;
}
